### ISSUE TITLE
docker run sometimes truncates stdout from container

### ISSUE THREAD

**Issue by mmm444** [docker/cli#4338]
### Description

We use a container to produce some data to stdout and redirect them to a file: `docker run --rm some_image > output_file` Recently we have found that the output_file is sometimes incomplete in the sense that some bytes from the end are missing.

The behaviour is not deterministic but it is more likely triggered by some output sizes than others.

When the truncation happens the dockerd log contains message like this:
```
time="2023-06-08T17:18:45.532479815Z" level=error msg="stream copy error: read /proc/self/fd/26: file already closed
```

I have a suspicion that it did not happen with the past versions since we use this mechanism for a long time (since 2016) and did not run into this issue before. The latest working version that I have access to is 20.10.21.

### Reproduce

We have been able to create a reproducer script that often triggers the truncation:
```bash
#!/bin/bash
for i in $(seq 20 27) ; do
  for j in -1 0 1 ; do
    num=$((2**i + j))
    docker run --rm alpine:3.18 sh -c "dd if=/dev/zero bs=$num count=1 2>/dev/null" > x
    res="$(stat -c%s x)"
    printf "expected: %9i actual: %9i ok: %i\n" $num $res $((num == res))
  done
done
```
It sends between 1M and 132M zeros over stdout and verifies that the received size matches the expected.

On our system it outputs for example:
```
expected:   1048575 actual:   1048575 ok: 1
expected:   1048576 actual:   1048576 ok: 1
expected:   1048577 actual:   1048577 ok: 1
expected:   2097151 actual:   2080768 ok: 0
expected:   2097152 actual:   2097152 ok: 1
expected:   2097153 actual:   2068480 ok: 0
expected:   4194303 actual:   4194303 ok: 1
expected:   4194304 actual:   4059136 ok: 0
expected:   4194305 actual:   4194305 ok: 1
expected:   8388607 actual:   8388607 ok: 1
expected:   8388608 actual:   8388608 ok: 1
expected:   8388609 actual:   8388609 ok: 1
expected:  16777215 actual:  16777215 ok: 1
expected:  16777216 actual:  16777216 ok: 1
expected:  16777217 actual:  16777217 ok: 1
expected:  33554431 actual:  33554431 ok: 1
expected:  33554432 actual:  33554432 ok: 1
expected:  33554433 actual:  33554433 ok: 1
expected:  67108863 actual:  67055616 ok: 0
expected:  67108864 actual:  67051520 ok: 0
expected:  67108865 actual:  67067904 ok: 0
expected: 134217727 actual: 134217727 ok: 1
expected: 134217728 actual: 134217728 ok: 1
expected: 134217729 actual: 134217729 ok: 1
```

### docker version

```bash
Client: Docker Engine - Community
 Version:           24.0.2
 API version:       1.43
 Go version:        go1.20.4
 Git commit:        cb74dfc
 Built:             Thu May 25 21:51:00 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          24.0.2
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.4
  Git commit:       659604f
  Built:            Thu May 25 21:51:00 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.21
  GitCommit:        3dce8eb055cbb6872793272b4f20ed16117344f8
 runc:
  Version:          1.1.7
  GitCommit:        v1.1.7-0-g860f061
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client: Docker Engine - Community
 Version:    24.0.2
 Context:    default
 Debug Mode: false

Server:
 Containers: 5
  Running: 3
  Paused: 0
  Stopped: 2
 Images: 9
 Server Version: 24.0.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 3dce8eb055cbb6872793272b4f20ed16117344f8
 runc version: v1.1.7-0-g860f061
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.15.0-73-generic
 Operating System: Ubuntu 22.04.2 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 15.46GiB
 Name: test
 ID: 64bed5d2-a573-4829-90ad-8b0a5764e2e3
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

### Additional Info

Hopefully this is the right part of the docker ecosystem to report this bug against. :)

**Comment by thaJeztah** [comment]
Thanks for reporting !This sounds like a possible duplicate of;

- https://github.com/moby/moby/issues/45689

Which is a bug in containerd, that will be fixed once this backport is merged and included in a containerd release;

- https://github.com/containerd/containerd/pull/8659

**Comment by laurazard** [comment]
It looks like it, yes! I've just rebased https://github.com/containerd/containerd/pull/8659, hopefully we can get it merged soon/a release so we can fix this over in moby.

**Comment by thaJeztah** [comment]
> Sorry about that.

Don't be! The linked issue was in another repository, and it can be very hard to find duplicates (I just recognised the issue, and knew there was another issue related to that). Thank you _so much_ for reporting the issue, and for verifying that the fix  will address your issue; it's appreciated 👍 🙌
